### PR TITLE
Change client/server version support

### PR DESCRIPTION
**Pull request by alexshtin**
**What changed?**
The way how client and server checks each other versions.

**Why?**
Previous schema was complicated and hard to understand/explain. This implementation follow simple rules:
1. All validation logic is on server. Client only sends headers. If header is missing assume that "client doesn't care" and no restriction is applied.
2. Client sends to server following headers:
    1. `client-name`: one of `temporal-go`, `temporal-java`, `temporal-cli`, or `temporal-server`.
    2. `client-version`: `0.30.0`
    3. `supported-server-versions`: `>1.1.1 <=1.4.0||^5.0.0` (https://github.com/blang/semver#ranges).
3. Server has `GetClusterInfo` API which returns its version and supported version ranges for every supported client.

**How did you test it?**
Modified existing tests.

**Potential risks**
No risks, because version check wasn't actually used and because missing headers are treated as "accepted" all combinations of old/new server and old/new clients should work.

### CODE REVIEW

File: common/headers/versionChecker.go [hunk 1]
```diff
@@ -28,134 +28,99 @@ import (
...
-	BaseFeaturesFeatureVersion = "1.0.0"
-
-	baseFeatures = "base-features"
+	ClientNameServer  = "temporal-server"
+	ClientNameGoSDK   = "temporal-go"
+	ClientNameJavaSDK = "temporal-java"
+	ClientNameCLI     = "temporal-cli"
+
+	ServerVersion = "0.31.0"
+	CLIVersion    = "0.31.0"
```
Comment (alexshtin) [inline common/headers/versionChecker.go:42]: These should be updated with every release.

File: common/headers/versionChecker.go [hunk 2]
```diff
@@ -28,134 +28,99 @@ import (
...
+	// SupportedServerVersions is used by CLI and inter role communication.
+	SupportedServerVersions = ">=0.31.0 <2.0.0"
+)
+
+var (
+	SupportedClients = map[string]string{
+		ClientNameGoSDK:   "<2.0.0",
+		ClientNameJavaSDK: "<2.0.0",
+		ClientNameCLI:     "<2.0.0",
+		ClientNameServer:  "<2.0.0",
```
Comment (alexshtin) [inline common/headers/versionChecker.go:53]: Assumes that version `2.0.0` will break compatibility.

File: common/persistence/dataInterfaces.go [hunk 3]
```diff
@@ -232,9 +232,6 @@ type (
 		CancelRequestId                        string
 		StickyTaskQueue                        string
 		StickyScheduleToStartTimeout           *time.Duration
-		ClientLibraryVersion                   string
```
Comment (mfateev) [inline common/persistence/dataInterfaces.go:235]: Why are we removing it? Looks like useful info. Obviously, field names should change.

File: common/metrics/runtime.go [hunk 4]
```diff
@@ -32,21 +32,22 @@ import (
...
 
-	// Branch is the VCS branch associated with this build.
-	Branch = "unknown"
+	// GitBranch is the git branch associated with this build.
+	GitBranch = "unknown"
 
-	// Version is the version associated with this build.
-	Version = "unknown"
+	// GitVersion is the git version (tag) associated with this build.
+	GitVersion = "unknown"
```
Comment (alexshtin) [inline common/metrics/runtime.go:50]: This version has an issue. If someone just `pull master` every time w/o fetching tags, this version becomes stale and never get updated.

File: proto/internal/temporal/server/api/adminservice/v1/request_response.proto [hunk 5]
```diff
@@ -161,8 +160,9 @@ message DescribeClusterRequest {
 }
 
 message DescribeClusterResponse {
-    temporal.api.version.v1.SupportedSDKVersions supported_client_versions = 1;
-    temporal.server.api.cluster.v1.MembershipInfo membership_info = 2;
+    map<string,string> supported_clients = 1;
+    string server_version = 2;
+    temporal.server.api.cluster.v1.MembershipInfo membership_info = 3;
```
Comment (alexshtin) [inline proto/internal/temporal/server/api/adminservice/v1/request_response.proto:165]: Breaking change.

File: proto/internal/temporal/server/api/persistenceblobs/v1/message.proto [hunk 6]
```diff
@@ -295,9 +295,9 @@ message WorkflowExecutionInfo {
     bytes event_branch_token = 45;
     int64 signal_count = 46;
     int64 history_size = 47 [deprecated = true];
-    string client_library_version = 48;
-    string client_feature_version = 49;
-    string client_impl = 50;
+    reserved 48;
+    reserved 49;
+    reserved 50;
```
Comment (alexshtin) [inline proto/internal/temporal/server/api/persistenceblobs/v1/message.proto:300]: This info is not persisted anymore as it is not used by code. @samarabbas, I am wondering if we still want to persist client name and version for debugging purpose.

File: common/headers/versionChecker.go [hunk 7]
```diff
@@ -28,134 +28,99 @@ import (
...
-	clientImpl := headers[1]
-
-	// If client doesn't provide version, it means it is fine with any server version (including current).
-	if clientVersion == "" {
-		return nil
+	headers := GetValues(ctx, ClientNameHeaderName, ClientVersionHeaderName, SupportedServerVersionsHeaderName)
+	clientName := headers[0]
+	clientVersion := headers[1]
+	supportedServerVersions := headers[2]
+
```
Comment (alexshtin) [inline common/headers/versionChecker.go:100]: Bellow is the core logic according to spec.

File: common/headers/headers.go [hunk 8]
```diff
@@ -31,29 +31,22 @@ import (
...
 )
 
 var (
 	versionHeaders = metadata.New(map[string]string{
-		ClientVersionHeaderName:        SupportedGoSDKVersion,
-		ClientFeatureVersionHeaderName: BaseFeaturesFeatureVersion,
-		ClientImplHeaderName:           GoSDK,
+		ClientNameHeaderName:              ClientNameServer,
+		ClientVersionHeaderName:           ServerVersion,
+		SupportedServerVersionsHeaderName: ">=0.31.0 <2.0.0",
```
[the remark this answers is not in the document]
Comment (alexshtin) [reply common/headers/headers.go]: I moved it to `const` and placed closer to each other. I don't want to do concatenation as I don't expect to change `SupportedServerVersions` very often. Not with every server release for sure. And `ServerVersion` will be changes with every release.

**Review by samarabbas — APPROVED** [review]

File: common/persistence/dataInterfaces.go (hunk 3 above)
Comment (alexshtin) [reply common/persistence/dataInterfaces.go:235]: After discussion with @samarabbas we decided to store client version info later in workflow history as version marker. Storing only last client version here doesn't make much sense.

**Comment by alexshtin** [comment]
Run manual backward compatibility tests: old canary -> new server, new canary -> old server. New canary -> new server also works :-).

**Review by mfateev — APPROVED** [review]

**Pull request merged**